### PR TITLE
[2.x] Improves console output

### PR DESCRIPTION
**Pull request by nunomaduro**
This pull request improves the output of tinker without breaking changes. Meaning that once merged, this pull request will be available even for Laravel 6 applications.

So, regarding the changes, here is some examples, the **before vs after**:

Screenshot 2022-07-27 at 14 00 01
 
Screenshot 2022-07-27 at 14 04 34

Screenshot 2022-07-27 at 14 07 00

### CODE REVIEW

File: src/Shell.php [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    public function formatException(Exception $exception): string
+    {
+        $message = $exception->getMessage();
+
+        if (Str::contains($message, 'psy/psysh/src/ExecutionLoopClosure.php')) {
+            [$message] = explode(' in /', $message);
+            [$message] = explode(' on line ', $message);
+        }
+
+        $severity = ($exception instanceof ErrorException) ? $this->getSeverity($exception) : 'error';
```
Comment (Aberdeener) [inline src/Shell.php:57]: Should we fully capitalize this severity text so it is consistent with “INFO” and “WARN”?

Comment (nunomaduro) [reply src/Shell.php:57]: Done.

**Comment by Wirone** [comment]
image

I understand UI improvements and the need to get good-looking output, but removing credits from the original tool that runs Tinker is unfortunate, at least. Tinker's repo contains 4 (FOUR) actual classes, the whole "fancy magic" is handled by [`PsySH`](https://github.com/bobthecow/psysh) which is on MIT licence - not only Laravel Tinker does not mention in any way that it's built on top of PsySH, it also now hides such info when running command. Personally, I find it really rude. This is not the way how you appreciate other people's work 😕

**Comment by nunomaduro** [comment]
@bobthecow Are you ok on having us removing the header from tinker? If not, we don't mind of adding it back.

File: src/Shell.php [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    {
+        return parent::doRun($input, $this->output = new SymfonyStyle($input, $output));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getHeader(): string
+    {
+        return '';
```
Comment (bobthecow) [inline src/Shell.php:42]: I'd prefer we keep the header. As @Wirone pointed out, this is the only indication to Tinker users that what they're using is actually (literally) 99% PsySH, with a bit of a shim to make it work in their app.

File: src/Shell.php [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function formatException(Exception $exception): string
+    {
+        $message = $exception->getMessage();
+
+        if (Str::contains($message, 'psy/psysh/src/ExecutionLoopClosure.php')) {
```
Comment (bobthecow) [inline src/Shell.php:52]: This one [should already be handled in PsySH](https://github.com/bobthecow/psysh/blob/main/src/Shell.php#L1214). Let's fix the bug there instead?

File: src/Shell.php [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+use Psy\Exception\Exception as PsyException;
+use Psy\Shell as BaseShell;
+use Symfony\Component\Console\Exception\InvalidArgumentException;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\Console\Style\SymfonyStyle;
+
+class Shell extends BaseShell
+{
+    const BUFF_PROMPT = '  ';
```
Comment (bobthecow) [inline src/Shell.php:19]: this exists for a reason :)

File: src/Shell.php [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+        }
+
+        if ($ret instanceof NoReturnValue) {
+            return;
+        }
+
+        $formatted = $this->presentValue($ret);
+
+        $message = collect(explode("\n", $formatted))->map(function ($line) {
+            return $this->grayExists() ? "<fg=gray;>│ </>$line" : "│ $line";
```
Comment (bobthecow) [inline src/Shell.php:104]: One reason multi-line output is indented but doesn't have any other characters is to make it possible to copy simple values out and use them without having to strip line numbers or whatever. Obviously this doesn't work with complex things like objects, but most scalars are "valid PHP" or close. Having a character at the start of every line of the output makes this more difficult.

**Review by bobthecow — COMMENTED** [review]
I love cleaner output!

I don't think I love extending `Psy\Shell` and doing string parsing to get it. It's a bit fragile, and it increases the overhead for making changes upstream because it'll impact Laravel users differently. 

It seems like most/all of this actually has nothing to do with Laravel, specifically, but is just … improvements to PsySH output? Can we move it all upstream and have the conversation there?

**Comment by nunomaduro** [comment]
@bobthecow Pull requesting this tomorrow to https://github.com/bobthecow/psysh.

File: src/Shell.php (hunk 2 above)
Comment (keradus) [reply src/Shell.php:42]: I believe we should apply such exposing of original tool not only for tinker/PsySH, but also for pint/PHP CS Fixer

Comment (keradus) [reply src/Shell.php:42]: And any other tools in similar case

**Pull request closed without merging**